### PR TITLE
[Feature] Add lossy cross-cast WASM bindings for scalar, integer, boolean, field, and group types

### LINKED ISSUE #1280: [Feature] Support Cast and Lossy Casts between Aleo Types
## 🚀 Feature

Developers have requested the ability to support cast operations between Aleo types.

## Motivation

These cast operations often are used to populate input data for functions, new records, are required for mapping access, and sometimes needed for custom cryptography done on the client side of Aleo dapps.

### PR DESCRIPTION
**Pull request by marshacb**
## Motivation

Aleo programs frequently need to cast between scalar types (field, group, scalar, boolean, integers). Currently the WASM layer doesn't expose these conversions, forcing SDK users to serialize to string, cross the WASM boundary, and re-parse, adding unnecessary overhead and complexity. This PR adds direct lossy cross-cast bindings (`toField`, `toGroup`, `toScalar`, `toBoolean`, `toU8`–`toI128`, `toAddress`) to all WASM scalar types, matching snarkVM's `cast_lossy` semantics. All conversions go through the field representation, with truncation for narrowing casts and mapping for field→group.

## Test Plan

- `yarn build:sdk` - TS compiles cleanly
- TS tests in `sdk/tests/cast-lossy.test.ts` covering:
  - **Field**: round-trip Field→Scalar→Field, Field→Group, Field→Address, Field→integers (U8–I128), Field→Boolean
  - **Integer**: cross-cast between integer types (U32→U8 truncation, U8→U32 widening), Integer→Field round-trip, Integer→Group, Integer→Scalar, Integer→Boolean, Integer→Address
  - **Boolean**: Boolean→Field (false=0, true=1), Boolean→integers, Boolean→Scalar, Boolean→Group, Boolean→Address
  - **Scalar**: Scalar→Field round-trip, Scalar→Group, Scalar→integers, Scalar→Boolean, Scalar→Address
  - **Group**: Group→Field, Group→Scalar, Group→integers, Group→Boolean, Group→Address, Address→Group round-trip
  - **Address**: Address→Field, Address→Group round-trip, Address→Scalar, Address→integers, Address→Boolean
- Existing `ExecutionRequest` error handling tests retained and passing
- Rust unit tests in `integer.rs` cover narrowing, widening, signed cross-casts, and `fromField`/`fromFieldLossy`

---

> [!NOTE]
> **Medium Risk**
> Adds many new WASM-exported conversion methods across core scalar types (Field/Scalar/Group/Boolean/integers/Address), which expands the public API and could introduce subtle semantic mismatches or edge-case bugs in lossy casting behavior. Changes are mostly additive and covered by new Rust and TS tests, reducing regression risk.
> 
> **Overview**
> Adds a comprehensive set of WASM bindings for *lossy and strict cross-type casts* between `Field`, `Scalar`, `Group`, `Boolean`, `Address`, and all integer types (`U8`–`U128`, `I8`–`I128`), generally routing through field representations and matching snarkVM `cast_lossy` semantics (including Elligator-2 fallback for `toGroupLossy`).
> 
> Introduces an integer cross-cast macro to generate `to*Lossy` methods for every integer-to-integer combination, plus new helpers like integer `toField`, `fromFieldLossy`, and `toBooleanLossy`, and adds coverage via a new `sdk/tests/cast-lossy.test.ts` suite and new Rust unit tests in `field.rs` and `integer.rs` (with a minor formatting fix in `sdk/tests/wasm.test.ts`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 18d8ac1ee80826ca9ab37db4fc174b1195978484. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: wasm/src/types/scalar.rs [hunk 1]
```diff
@@ -149,6 +174,92 @@ impl Scalar {
...
+        // Safe: Scalar::to_field() is infallible for all valid scalar values.
+        let field: FieldNative = self.0.to_field().unwrap();
+        let group: GroupNative = field.cast_lossy();
+        Group::from(group)
+    }
+
+    /// Cast the scalar to an Address (via Group).
+    #[wasm_bindgen(js_name = "toAddress")]
+    pub fn to_address(&self) -> Address {
+        Address::from_group(self.to_group())
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/scalar.rs]: AFAIU `Scalar::to_field()` is infallible for all valid scalar values so a `Scalar` can only be constructed through validated paths. The `// Safe:` comment documents this. Changing the return type to Result would worsen the API for callers with no practical safety benefit.

File: wasm/src/types/scalar.rs [hunk 2]
```diff
@@ -149,6 +174,92 @@ impl Scalar {
...
+    /// Cast the scalar to a U16 with lossy truncation.
+    #[wasm_bindgen(js_name = "toU16")]
+    pub fn to_u16(&self) -> U16 {
+        U16::from(U16Native::from_field_lossy(&self.0.to_field().unwrap()))
+    }
+
+    /// Cast the scalar to a U32 with lossy truncation.
+    #[wasm_bindgen(js_name = "toU32")]
+    pub fn to_u32(&self) -> U32 {
+        U32::from(U32Native::from_field_lossy(&self.0.to_field().unwrap()))
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/scalar.rs]: same as above

File: wasm/src/types/integer.rs [hunk 3]
```diff
@@ -205,6 +213,41 @@ macro_rules! impl_integer {
...
+                // Safe: Integer::to_field() is infallible for all valid integer values.
+                let field = self.0.to_field().unwrap();
+                let group: GroupNative = field.cast_lossy();
+                Group::from(group)
+            }
+
+            /// Cast the integer to an Address (via Group).
+            #[wasm_bindgen(js_name = "toAddress")]
+            pub fn to_address(&self) -> Address {
+                Address::from_group(self.to_group())
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/integer.rs]: same as above

File: wasm/src/types/integer.rs [hunk 4]
```diff
@@ -253,3 +296,174 @@ impl_integer!(U16, U16Native);
...
+            /// Cast to I64 with lossy truncation.
+            #[wasm_bindgen(js_name = "toI64")]
+            pub fn to_i64(&self) -> I64 {
+                I64::from(I64Native::from_field_lossy(&self.0.to_field().unwrap()))
+            }
+
+            /// Cast to I128 with lossy truncation.
+            #[wasm_bindgen(js_name = "toI128")]
+            pub fn to_i128(&self) -> I128 {
+                I128::from(I128Native::from_field_lossy(&self.0.to_field().unwrap()))
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/integer.rs]: same as above

File: wasm/src/types/boolean.rs [hunk 5]
```diff
@@ -138,6 +165,101 @@ impl Boolean {
...
+        let bits = vec![*self.0];
+        Field::from(FieldNative::from_bits_le(&bits).unwrap())
+    }
+
+    /// Cast the boolean to a Scalar element (false=0, true=1).
+    #[wasm_bindgen(js_name = "toScalar")]
+    pub fn to_scalar(&self) -> Scalar {
+        // Safe: from_bits_le with a single bit is infallible.
+        let bits = vec![*self.0];
+        Scalar::from(ScalarNative::from_bits_le(&bits).unwrap())
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/boolean.rs]: Applied the point on the `vec!` to slice optimization.

File: wasm/src/types/integer.rs [hunk 6]
```diff
@@ -205,6 +213,41 @@ macro_rules! impl_integer {
...
+            }
+
+            /// Cast the integer to a Group element (via Field, Elligator-2 fallback).
+            #[wasm_bindgen(js_name = "toGroup")]
+            pub fn to_group(&self) -> Group {
+                // Safe: Integer::to_field() is infallible for all valid integer values.
+                let field = self.0.to_field().unwrap();
+                let group: GroupNative = field.cast_lossy();
+                Group::from(group)
+            }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/integer.rs]: This one is redundant.

```suggestion
```

File: wasm/src/types/scalar.rs [hunk 7]
```diff
@@ -149,6 +174,92 @@ impl Scalar {
...
+    #[wasm_bindgen(js_name = "toBoolean")]
+    pub fn to_boolean(&self) -> Boolean {
+        Boolean::new(self.0.to_bits_le()[0])
+    }
+
+    /// Cast the scalar to a Group element (via Field, then Elligator-2 fallback).
+    #[wasm_bindgen(js_name = "toGroup")]
+    pub fn to_group(&self) -> Group {
+        // Safe: Scalar::to_field() is infallible for all valid scalar values.
+        let field: FieldNative = self.0.to_field().unwrap();
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/scalar.rs]:
```suggestion
        let field: FieldNative = self.0.to_field().map_err(|e| e.to_string())?;
```

File: wasm/src/types/field.rs [hunk 8]
```diff
@@ -152,6 +179,99 @@ impl Field {
...
+    /// Cast the field element to a Group element.
+    ///
+    /// Uses the snarkVM cast_lossy path: tries x-coordinate recovery first,
+    /// falls back to the generator for field == 1, and applies Elligator-2
+    /// otherwise. This conversion never fails.
+    #[wasm_bindgen(js_name = "toGroup")]
+    pub fn to_group(&self) -> Group {
+        let group: GroupNative = self.0.cast_lossy();
+        Group::from(group)
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/field.rs]: There should be two methods

1. toGroupLossy which performs this behavior
2. toGroup which performs the strict behavior

**Review by iamalwaysuncomfortable — CHANGES_REQUESTED** [review]
Overall really great but a couple of key things

1. We should be explicit about the casting here, on some of the types there's a safe cast (often fallible) and a lossy cast and anything that is lossy should have the suffix `-Lossy` and anything that isn't a lossy cast can just have the regular naming here (i.e. `toGroup` (safe cast) `toGroupLossy` (lossy cast)). 
2. Turn any .unwraps() into fallible errors.
3. Also on non-integer types, we can remove the self cast methods (`Group -> Group`, `Scalar -> Scalar`, `Field -> Field`)

File: wasm/src/account/address.rs [hunk 9]
```diff
@@ -218,6 +237,88 @@ impl Address {
...
         false
     }
+
+    // ── cast conversions ───────────────────────────────────────────────
+
+    /// Cast the address to a Field element with lossy truncation (x-coordinate of the underlying group point).
+    #[wasm_bindgen(js_name = "toFieldLossy")]
+    pub fn to_field_lossy(&self) -> Field {
+        Field::from(self.0.to_group().to_x_coordinate())
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/account/address.rs]: This is just the regular cast to a field not the lossy, I don't believe there is a lossy cast from Address/Group --> Field.


```suggestion
    /// Cast the address to a Field element with lossy truncation (x-coordinate of the underlying group point).
    #[wasm_bindgen(js_name = "toField")]
    pub fn to_field_lossy(&self) -> Field {
        Field::from(self.0.to_group().to_x_coordinate())
    }
```

File: wasm/src/types/field.rs [hunk 10]
```diff
@@ -152,6 +179,108 @@ impl Field {
...
+    #[wasm_bindgen(js_name = "toScalarLossy")]
+    pub fn to_scalar_lossy(&self) -> Scalar {
+        Scalar::from(ScalarNative::from_field_lossy(&self.0))
+    }
+
+    /// Cast the field element to a Boolean with lossy truncation (extracts least-significant bit).
+    #[wasm_bindgen(js_name = "toBooleanLossy")]
+    pub fn to_boolean_lossy(&self) -> Boolean {
+        Boolean::new(self.0.to_bits_le()[0])
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/field.rs:208]: We should include the regular toBoolean as well here - [ref](https://github.com/ProvableHQ/snarkVM/blob/staging/console/program/src/data/literal/cast/field.rs#L31-L45)

File: wasm/src/types/field.rs [hunk 11]
```diff
@@ -152,6 +179,108 @@ impl Field {
...
+    pub fn to_group_lossy(&self) -> Group {
+        let group: GroupNative = self.0.cast_lossy();
+        Group::from(group)
+    }
+
+    /// Cast the field element to an Address with lossy conversion (via Group).
+    #[wasm_bindgen(js_name = "toAddressLossy")]
+    pub fn to_address_lossy(&self) -> Address {
+        Address::from_group(self.to_group_lossy())
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/field.rs]: We should add the regular `toAddress` as well (i.e. wrapping an address over the to_group method above).

File: wasm/src/types/scalar.rs [hunk 12]
```diff
@@ -149,6 +174,90 @@ impl Scalar {
...
         self.0 == ScalarNative::from(other)
     }
+
+    // ── cast conversions ───────────────────────────────────────────────
+
+    /// Cast the scalar to a Boolean with lossy truncation (extracts least-significant bit).
+    #[wasm_bindgen(js_name = "toBooleanLossy")]
+    pub fn to_boolean_lossy(&self) -> Boolean {
+        Boolean::new(self.0.to_bits_le()[0])
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/scalar.rs:197]: We should add the regular toBoolean here as well (in addition to this method) - [ref](https://github.com/ProvableHQ/snarkVM/blob/staging/console/program/src/data/literal/cast/scalar.rs#L33-L47)

File: wasm/src/types/scalar.rs [hunk 13]
```diff
@@ -149,6 +174,90 @@ impl Scalar {
...
+        let field: FieldNative = self.0.to_field().map_err(|e| e.to_string())?;
+        let group: GroupNative = field.cast_lossy();
+        Ok(Group::from(group))
+    }
+
+    /// Cast the scalar to an Address with lossy conversion (via Group).
+    #[wasm_bindgen(js_name = "toAddressLossy")]
+    pub fn to_address_lossy(&self) -> Result<Address, String> {
+        Ok(Address::from_group(self.to_group_lossy()?))
+    }
```
Comment (iamalwaysuncomfortable) [inline wasm/src/types/scalar.rs]: We should support the regular non-lossy toAddress here as well - [ref](https://github.com/ProvableHQ/snarkVM/blob/staging/console/program/src/data/literal/cast/scalar.rs#L27-L30)

(Basically, an address wrapping a regular non-lossy group cast)

**Review by iamalwaysuncomfortable — COMMENTED** [review]
A couple more changes to request, almost there.

File: wasm/src/account/address.rs (hunk 9 above)
Comment (marshacb) [reply wasm/src/account/address.rs]: Good call. Group/Address to Field is just x-coordinate extraction, there's only one path and it's deterministic. Renamed to toField on both Address and Group.

**Review by iamalwaysuncomfortable — APPROVED** [review]

File: wasm/src/types/boolean.rs [hunk 14]
```diff
@@ -138,6 +165,111 @@ impl Boolean {
...
+    }
+
+    /// Cast the boolean to a Group element (lossy, via Field with Elligator-2 fallback).
+    /// This conversion never fails.
+    #[wasm_bindgen(js_name = "toGroupLossy")]
+    pub fn to_group_lossy(&self) -> Result<Group, String> {
+        let field = FieldNative::from_bits_le(&[*self.0]).map_err(|e| e.to_string())?;
+        let group: GroupNative = field.cast_lossy();
+        Ok(Group::from(group))
+    }
```
[the remark this answers is not in the document]
Comment (marshacb) [reply wasm/src/types/boolean.rs:198]: Leaving as is for now to keep consistent with having no unwraps.
